_posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md

```diff
@@ -0,0 +1,54 @@
+
+This is where Kani comes in very handy. From its definition, Durations have a minimum and maximum value. On an unsigned 64 bit integer, we can store enough nanoseconds for four full centuries a bit. But we want want normalize all operations such that the nanosecond counter stays within one century (until we're reached the maximum number of centuries). Centuries are stored in a _signed_ 16 bit integer, and are the only field that store whether the duration is positive or negative.
+
+At first sight, it seems like a relatively simple task to make sure that this math is correct. It turns out that lots of bug may happen near the min and max durations, and when performing operations between very large and very small durations. Kani has helped fix at least eight different categories of bugs in a single pull request: <https://github.com/nyx-space/hifitime/pull/192>. Most of these were bugs near the boundaries of a Duration definition: around the zero, maximum, and minimum durations. But many of the bugs were on important and common operations: partial equality, negation, addition, and subtraction operations.
+
+One of the great things of Kani is that it doesn't require explicitly exhaustive tests. In fact, Kani operates on the mid-layer representation. It then solves for satifiability: "can this operation ever be executed, and if so, what are the cases where it may cause issues."
+
+Thanks to how Kani analyzes a program, tests can either be bounded or not. A bounded test includes an assertion: execute a set of instructions and then check something. This is a typical test case.
+
+Kani can also have unbounded tests where there is no explicit condition to check. Instead, only the successive operations of a function call are executed, and each are tested by Kani for failure cases by analyzing the inputs and finding cases where inputs will lead to runtime errors like overflows. This approach is how most of the bugs in hifitime have been found.
+
+Unbounded tests effectively ensure that sanity of the operations in a given function call. Bounded tests provide the same while also checking equalities. If either of these tests fail, Kani can provide a test failure report outlining the sequence of operations, and the binary representation of each intermediate operation, to help the developer gain an understanding of why their implementation is incorrect.
+
+The overhead to implement tests in Kani is very low, and the benefits are immense. Basically, write a Kani verification like a unit test, run the model verifier, and you've formally verified this part of the code. Amazing!
```

Comment:
We might want to reconsider the use of "tests" in these paragraphs.

Reply:
It seems like the way bounded/unbounded distinction is made here conflicts with the usual definitions.

Reply:
It's quite likely that I don't understand the difference between bounded and unbounded tests: can you explain the difference ?